### ISSUE TITLE
need promregator.targets[].port to define specific port 8080 to fetch

### ISSUE THREAD

**Issue by ManuPin** [promregator/promregator#190]
## Environment
I deploy promregator in a space of cloud foundry with versionning and manifest : 
---
applications:
  - name: promregator
    instances: 1
    memory: 1G
    disk_quota: 512M
    path: public/promregator-0.8.2.jar
    buildpacks:
      - java_buildpack_v4_35
    env:
      SPRING_APPLICATION_JSON: '{"promregator":{"authenticator":{"type":"none"},"targets":[{"orgName":"orgName","spaceName":"spaceName","path":"/actuator/prometheus","protocol":"http"}]},"cf":{"api_host":"api.cloudfoundry","username":"","password":""}}'

The api version:    2.147.0, And promeregator discover the app but can't fetch metric because can't find promregator.targets[].port parameter

## Expected Behavior
<!-- describe what behavior i woul expect a key value to define the port to fecth,in Subgroup "promregator.targets" , a promregator.targets[].port. Then we can define , protocol, path and port to fetch prometheus metrics for each target.

**Comment by eaglerainbow** [comment]
Hi @ManuPin,

Thanks for joining in here! And thanks reporting your issue.

I am trying to understand your situation & context, but honestly I have a little trouble with. Here is what I got: 
* You are deploying Promregator to Cloudfoundry.
* You don't specify any routes for it (or just use the default route provided)
* Configuration happens using environment variable `SPRING_APPLICATION_JSON`. In that configuration you are defining a space to scrape. You want to make use of the HTTP protocol.
* Apparently, you intend to provide port 8080 within the configuration.

Note that there is no such thing as a `port` parameter to a target (see also https://github.com/promregator/promregator/blob/af3893bb2090fe10089ed1d14ddd43ebe62dabf0/docs/config.md#subgroup-promregatortargets ) -- and that is intended!
To explain that keep our [architecture picture](https://github.com/promregator/promregator/blob/af3893bb2090fe10089ed1d14ddd43ebe62dabf0/docs/architecture.png) in mind: Promregator will always be querying metrics going through the (go)routers of Cloud Foundry! Even if you deploy Promregator to CF itself, it does not make use of internal network communication. That is why I don't get the hang of it why you want to specify port 8080 in the configuration. Specifying it would mean that you try to connect to the haproxies of CF (in front of the gorouters) using port 8080... Promregator does not talk to your applications directly...

Does this already help you and even may answer your issue? Or do you have some different approach on your mind which I couldn't read out of your question initially? If it is the latter, please elaborate on your use case a little more such that we are able here to comprehend. Thanks!

Merry Christmas!

**Comment by ScottGuymer** [comment]
I think I get what the OP is asking for and have a similar need.

- deploy apps via docker images to CF
- deploy promregator to same CF instance
- each app docker image exposes 2 ports 
  - 8080 for app traffic
  - 9090 for metrics only
- use public route for app traffic bound to 8080
- internal route for metrics traffic (want to keep metrics private)
- use internal routes between promregator and metrics endpoint
- internal route does not map ports and requires you to define a port
- internal route has a network-policy that defines access between promregator and app on port 9090

So in our case we have an app called `app-a` which has in internal route defined `app-a-metrics.apps.internal`

We configure promregator to use the internal route via `preferredRouteRegex=".*-metrics"` and use the http protocol.

When making a request promregator then uses `app-a-metrics.apps.internal:80`

There is no network policy for this (and the app is not listening) so the request fails.

The request needs to be made on `app-a-metrics.apps.internal:8080` to be successful.

tl;dr; need to be able to set the port that the scrape is performed on.

**Comment by checketts** [comment]
@ScottGuymer When using a internal route, is there a method to call a specific app instance? One of Promregator's roles, is to be a proxy to the correct instance so the metrics are collected correctly. The Go Router handles that routing for internal routes, how is it handled?

If I recall, perhaps the app instance is also prefixed? So the route would be `0.app-a-metrics.apps.internal:8080`? Could you please verify that?

**Comment by ScottGuymer** [comment]
It appears you are correct 

(500 in this case means its speaking to the app)

```
~ # wget my-app.apps.internal:8080
Connecting to my-app.apps.internal:8080 (10.254.165.176:8080)
wget: server returned error: HTTP/1.1 500 Internal Server Error

~ # wget 0.my-app.apps.internal:8080
Connecting to 0.my-app.apps.internal:8080 (10.255.20.75:8080)
wget: server returned error: HTTP/1.1 500 Internal Server Error

~ # wget 1.my-app.apps.internal:8080
Connecting to 1.my-app.apps.internal:8080 (10.254.165.176:8080)
wget: server returned error: HTTP/1.1 500 Internal Server Error

~ # wget 2.my-app.apps.internal:8080
wget: bad address '2.my-app.apps.internal:8080'
```

Im not sure how I would effectively test if the header to select instance works.

**Comment by ScottGuymer** [comment]
I took a go at adding some support for this with a little more automation that adding extra config.

#194 

Its a work in progress and feedback would be appreciated.

**Comment by eaglerainbow** [comment]
I don't know whether with #194 we are on the right track: Mainly promregator is nothing else than a lot of stuff written around the concept of going through the gorouter. It shall send an appropriately-formatted request to it, which takes care that gorouter will open a tunnel to the corresponding application container. Promregator exists, because formulating it with means Prometheus provides, this is not possible.
If you wanted to exaggerate a little, you could also say that Promregator's heart is [one single line of code](https://github.com/promregator/promregator/blob/bbe3aa3c21caf8bbc5a7f9352794bd54910ab46a/src/main/java/org/cloudfoundry/promregator/fetcher/CFMetricsFetcher.java#L125) - and all the other stuff is just a whole bunch of fuzz around it to make that line fly. 
The same aspect could also be seen at our basic architecture picture:

![Architecture of Promregator](https://github.com/promregator/promregator/blob/bbe3aa3c21caf8bbc5a7f9352794bd54910ab46a/src/main/java/org/cloudfoundry/promregator/fetcher/CFMetricsFetcher.java#L125)

Besides the fact that all scraping requests will go through gorouter, the diagram outlines also another fundamental assumption of promregator's concept: Promregator *may be* operated outside of the targeted Cloud Foundry landscape. There are folks out there, which run promregator on CF as well (which I would not reject as a whole), but that is a special, own decision - and in fact, this never was a design principle for promregator. To the contrary: In the beginning I was assuming that promregator was run "very near" to prometheus (there were even times where I had considered promregator to be a plugin to prometheus) - and hence somewhere in your local intranet. That is also why there wasn't any authentication scheme implemented for promregator's endpoints in the beginning.

The approach of #194, however, breaks with some fundamental assumptions now:

* Promregator *must* be installed as CF application on the same CF landscape as the applications.
* It will *never* send its requests through gorouter anymore (no "tunnel" anymore), but will directly query metrics from the application "through the backyard". 

Or to put it into harsh words: We drain promregator from [its origin](https://github.com/promregator/promregator/blob/bbe3aa3c21caf8bbc5a7f9352794bd54910ab46a/src/main/java/org/cloudfoundry/promregator/fetcher/CFMetricsFetcher.java#L125).

That is why I currently have the impression that we are making a major (but rather common software development) mistake: Because with some second scenario, for which the software *never* was designed to be used, we see "overlap" (in particular the metadata engine to determine what applications need to be scraped). We therefore try to "reuse existing software for a different purpose". In such cases you typically think that it is "just a couple of lines here and there to change", but executing on it would put promregator into a different context - and with that all our design and development assumptions *may* be invalidated (and we might not see that).  Doing so is a very dangerous act - and usually a doubtfully-joyful source of bugs and troubles. 

tl&dr; I don't think that this (#194) is promregator anymore - but more *something else*. I don't know what exactly it is. Yet, I don't want to discourage you from doing it in general. But eventually, I think, we shouldn't merge it into this repository's `master` codeline.

**Comment by ScottGuymer** [comment]
I totally understand where you are coming from and how this change may diverge a little from the original architecture. 

IMO it does push more into the service discovery for Prometheus in CF. As I think the proxying of the scraping would no longer be strictly needed when using internal routing.

I have managed to make the code all "automatic" so that it can identify if you are using an internal route for scraping and format the request URL correctly. this means both internal and external scraping could work in the same deployment of promregator. Of course, there are a few obstacles in the way such as network policies.

I do think that both the usecase for internal and external scaping are at least adjacent and could either be done in promregator or in "something else". There could even be use cases where both are needed as part of the same solution.

I'm happy to take your lead on this and either continue to work on adding to promregator or, with your agreement, fork and change the scope of the project into something focussed on internal service discovery.

**Comment by eaglerainbow** [comment]
> There could even be use cases where both are needed as part of the same solution.

Mmmh... there you have a point. This would clearly speak against separation. 

> I'm happy to take your lead on this [...]

The point here is, that I don't have a total clear opinion about it yet. My temporary suggestion therefore would be that we continue with it as it is right now: Go ahead with your PR - determine what is necessary to make things fly. We should also optionally toggle writing the header - and perhaps put things on better conceptual feet, i.e. by implementing an own [MetricsFetcher](https://github.com/promregator/promregator/blob/bbe3aa3c21caf8bbc5a7f9352794bd54910ab46a/src/main/java/org/cloudfoundry/promregator/fetcher/MetricsFetcher.java). Anyhow, I am not so super-happy about that piece of implementation (as it is not asynchronous, but blocking). 
Once we see in your PR how much you had to diverge from our mainline, let's revisit this topic again. I am open to discuss all options - amongst them are:
* Integrate it as a new feature (which might imply that we have to bump a major version) with manual switching
* Integrate it as a new feature (which is toggled automatically) (with which still I have some doubts)
* Split Promregator and outsource the service discovery part to a library which permits reuse in "classic Promregator" and your "something else" (which maybe is just another repository in the same github organization)
* Fork Promregator and let it consciously diverge (hey, it's open source after all!)

WDYT?

**Comment by ScottGuymer** [comment]
I think thats fair. I will complete the work for the PR to integrate the internal routing. Now i have played with it more I think it's possible to do without breaking any existing use cases. I think this is the case because an internal use case would never have worked before anyway.

There is also a choice to make if the promregator continues to serve as a proxy for internal requests or simply provides the service discovery and allows prometheus to scrape directly as it is fully capable of doing so without the header functionality added by promregator.

**Comment by eaglerainbow** [comment]
> and allows prometheus to scrape directly as it is fully capable of doing so without the header functionality added by promregator.

That statement brings us to a very interesting aspect - and may be even the key to everything: We would need to know where prometheus (our consumer) is located, i.e. whether it is part of the platform (to be able to access the internal hostnames) or not. 
What about the idea to add another (optional) parameter to the service discovery endpoint, which explicitly tells us so. If the parameter was not set, we can already filter out all internal routes altogether, as the consumer cannot access them anyway.
If the parameter was set, we may put a preference on internal routes and return the internal hostname directly as target, bypassing promregator in its reverse proxy function. 
How does that sound?

**Comment by ScottGuymer** [comment]
Yeah, I think we can put config and filtering around it where possible.

The internal usages would not have worked in either case now as it was missing the port number config. So, it is unlikely we will be breaking anyone's current functionality.

I think a switch would be useful to allow promregator to either act as a proxy (or not) for internal routes only, leaving other current functionality intact. 

Making it prefer internal routes is already trivial with adding a route regex, we could make it apply this automatically but would have to change the default behaviour of selecting the first route.

It would then be possible to build future functionality on what I add in here.

**Comment by ScottGuymer** [comment]
I'm getting #195 to a good place now where it needs to be reviewed to make sure I'm on the right track and have captured all of the potential problems.

I have added the functionality needed to detect if the selected route is an internal route or not and to construct the correct URL for the internal route.

**Comment by eaglerainbow** [comment]
sure thing. Will have a look at it. However, it'll be Wednesday evening at the earliest before I will get to it (too many other things today and tomorrow).
Will update here...

**Comment by eaglerainbow** [comment]
@ScottGuymer Now that #195  is merged, I have triggered the build for the release candidate. However, I see a lot of updates in the build server's queue. So, it will take a while until our rc version is built. Plus, I'll be off for some hours soon. Hence, I will ping you again, once everything we need is in place.

**Comment by eaglerainbow** [comment]
I just released [v0.9.0-rc1](https://github.com/promregator/promregator/releases/tag/v0.9.0-rc1). 
@tlkamp Please feel free to test the release candidate in your environment. @ScottGuymer and I have altered the original approach significantly such that a lot less roundtrips to the CFCC are necessary (compared to the original idea). Yet, we still have increased them. Therefore, testing is still appreciated.

**Comment by ScottGuymer** [comment]
I have been deploying and testing this out on my infra and internal routing now works a treat for me.

I am not doing any external scraping so i cant verify that it works in that case at the moment,

**Comment by ScottGuymer** [comment]
Still no issues here.

As a follow up or I am thinking about adding the option for making the discovery endpoint return the direct URL rather than acting as a proxy as it would make it easier to see which targets are up or down.

**Comment by tlkamp** [comment]
@eaglerainbow @ScottGuymer sorry, but I recently left the employer that was using Promregator internally. I will unfortunately not be able to run a test against that environment.

**Comment by eaglerainbow** [comment]
[tlkamp]
> sorry, but I recently left the employer that was using Promregator internally.

Many thanks for your transparency! Unfortunate to hear that...

> I will unfortunately not be able to run a test against that environment.

That's okay. We will take it from here...

[ScottHuymer]
> Still no issues here.

Yeah, same at my premise: Running in a test environment for a several days now, but we can't spot any difference in behavior (fortunately).

> As a follow up or I am thinking about adding the option for making the discovery endpoint return the direct URL rather than acting as a proxy as it would make it easier to see which targets are up or down.

For sure, give it a try. But note that this would change the model yet again: Running on container-2-container communication on CF, this would imply that also your Prometheus would have to be located in the same network as your containers. If that was the case, I am asking myself how much benefit Promregator still would give to you: If you are so close to your application's containers, why can't you then generate the scraping configuration file yourself? All what Promregator then is still doing, is a little bit of discovery through the CF API once in a while...

**Comment by ScottGuymer** [comment]
Yeah, it makes promregator more of a service discovery tool for CF rather than a proxy. 

It depends how you look at it. If your use case covers both internal and external targets to scrape, then it could be a useful setup to have.

Writing the code that promregator already has for SD from scratch would be a fair bit of work (of course not impossible). But I think that having this logic together in one tool that can do both is better than different implementations.

I am looking at eventually having the use case for both internal scraping in multiple CF instances coupled with then scraping multiple Prometheus instances into a single global view Prometheus and this would have to be done using internal and external scraping together. It would be great to have one tool to accomplish this .
